File: app/src/main/java/remix/myplayer/service/notification/NotifyImpl.java
package remix.myplayer.service.notification;

import android.app.Notification;
import android.app.PendingIntent;
import android.content.Context;
import android.graphics.Bitmap;
import android.os.Build;
import android.support.v4.app.NotificationCompat;
import android.view.View;
import android.widget.RemoteViews;

import remix.myplayer.R;
import remix.myplayer.bean.mp3.Song;
import remix.myplayer.request.RequestConfig;
import remix.myplayer.request.network.RemoteUriRequest;
import remix.myplayer.service.MusicService;
import remix.myplayer.util.ColorUtil;
import remix.myplayer.util.Constants;
import remix.myplayer.util.DensityUtil;
import remix.myplayer.util.LogUtil;
import remix.myplayer.util.SPUtil;

import static remix.myplayer.service.MusicService.copy;
import static remix.myplayer.util.ImageUriUtil.getSearchRequestWithAlbumType;

/**
 * Created by Remix on 2017/11/22.
 */

public class NotifyImpl extends Notify {
    private RemoteViews mRemoteView;
    private RemoteViews mRemoteBigView;

    public NotifyImpl(MusicService context) {
        super(context);
    }

    @Override
    public void updateForPlaying() {
        mRemoteBigView = new RemoteViews(mService.getPackageName(),R.layout.notification_big);
        mRemoteView = new RemoteViews(mService.getPackageName(),R.layout.notification);
        boolean isPlay = MusicService.isPlay();

        buildAction(mService);
        Notification notification = buildNotification(mService);

        if((MusicService.getCurrentMP3() != null)) {
            boolean isSystemColor = SPUtil.getValue(mService,SPUtil.SETTING_KEY.SETTING_NAME,SPUtil.SETTING_KEY.NOTIFY_SYSTEM_COLOR,Build.VERSION.SDK_INT < Build.VERSION_CODES.N);

            Song song = MusicService.getCurrentMP3();
            //设置歌手，歌曲名
            mRemoteBigView.setTextViewText(R.id.notify_song, song.getTitle());
            mRemoteBigView.setTextViewText(R.id.notify_artist_album, song.getArtist() + " - " + song.getAlbum());

            mRemoteView.setTextViewText(R.id.notify_song,song.getTitle());
            mRemoteView.setTextViewText(R.id.notify_artist_album,song.getArtist() + " - " + song.getAlbum());

            //设置了黑色背景
            if(!isSystemColor){
                mRemoteBigView.setTextColor(R.id.notify_song, ColorUtil.getColor(R.color.night_textcolor_primary));
                mRemoteView.setTextColor(R.id.notify_song, ColorUtil.getColor(R.color.night_textcolor_primary));
                //背景
                mRemoteBigView.setImageViewResource(R.id.notify_bg, R.drawable.bg_notification_black);
                mRemoteBigView.setViewVisibility(R.id.notify_bg, View.VISIBLE);
                mRemoteView.setImageViewResource(R.id.notify_bg, R.drawable.bg_notification_black);
                mRemoteView.setViewVisibility(R.id.notify_bg,View.VISIBLE);
            }
            //设置播放按钮
            if(!isPlay){
                mRemoteBigView.setImageViewResource(R.id.notify_play, R.drawable.notify_play);
                mRemoteView.setImageViewResource(R.id.notify_play, R.drawable.notify_play);
            }else{
                mRemoteBigView.setImageViewResource(R.id.notify_play, R.drawable.notify_pause);
                mRemoteView.setImageViewResource(R.id.notify_play, R.drawable.notify_pause);
            }
            //设置封面
            final int size = DensityUtil.dip2px(mService,128);

            new RemoteUriRequest(getSearchRequestWithAlbumType(song),new RequestConfig.Builder(size,size).build()){
                @Override
                public void onError(String errMsg) {
                    mRemoteBigView.setImageViewResource(R.id.notify_image, R.drawable.album_empty_bg_day);
                    mRemoteView.setImageViewResource(R.id.notify_image, R.drawable.album_empty_bg_day);
                    pushNotify(notification);
                }

                @Override
                public void onSuccess(Bitmap bitmap) {
                    try {
                        Bitmap result = copy(bitmap);
                        if(result != null) {
                            mRemoteBigView.setImageViewBitmap(R.id.notify_image, result);
                            mRemoteView.setImageViewBitmap(R.id.notify_image,result);
                        } else {
                            mRemoteBigView.setImageViewResource(R.id.notify_image, R.drawable.album_empty_bg_day);
                            mRemoteView.setImageViewResource(R.id.notify_image, R.drawable.album_empty_bg_day);
                        }
                    } catch (Exception e){
                        LogUtil.d("Notify",e.toString());
                    } finally {
                        pushNotify(notification);
                    }
                }

            }.load();
        }
    }

    private Notification buildNotification(Context context) {
        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, PLAYING_NOTIFICATION_CHANNEL_ID);
        builder.setContent(mRemoteView)
                .setCustomBigContentView(mRemoteBigView)
                .setContentText("")
                .setContentTitle("")
                .setShowWhen(false)
                .setPriority(NotificationCompat.PRIORITY_MAX)
                .setOngoing(MusicService.isPlay())
                .setContentIntent(getContentIntent())
                .setSmallIcon(R.drawable.notifbar_icon);
        builder.setCustomBigContentView(mRemoteBigView);
        builder.setCustomContentView(mRemoteView);
        return builder.build();
    }

    private void buildAction(Context context) {
        //添加Action
        //切换
        PendingIntent playIntent = buildPendingIntent(context, Constants.TOGGLE);
        mRemoteBigView.setOnClickPendingIntent(R.id.notify_play,playIntent);
        mRemoteView.setOnClickPendingIntent(R.id.notify_play,playIntent);
        //下一首
        PendingIntent nextIntent = buildPendingIntent(context, Constants.NEXT);
        mRemoteBigView.setOnClickPendingIntent(R.id.notify_next, nextIntent);
        mRemoteView.setOnClickPendingIntent(R.id.notify_next, nextIntent);
        //上一首
        PendingIntent prevIntent = buildPendingIntent(context, Constants.PREV);
        mRemoteBigView.setOnClickPendingIntent(R.id.notify_prev,prevIntent);

        //关闭通知栏
        PendingIntent closeIntent = buildPendingIntent(context,Constants.CLOSE_NOTIFY);
        mRemoteBigView.setOnClickPendingIntent(R.id.notify_close, closeIntent);
        mRemoteView.setOnClickPendingIntent(R.id.notify_close,closeIntent);
    }

}

File: app/src/main/java/remix/myplayer/service/notification/NotifyImpl.kt
package remix.myplayer.service.notification

import android.app.Notification
import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.view.View
import android.widget.RemoteViews
import androidx.core.app.NotificationCompat
import com.bumptech.glide.request.target.CustomTarget
import com.bumptech.glide.request.transition.Transition
import com.bumptech.glide.signature.ObjectKey
import remix.myplayer.R
import remix.myplayer.glide.GlideApp
import remix.myplayer.glide.UriFetcher
import remix.myplayer.service.Command
import remix.myplayer.service.MusicService
import remix.myplayer.util.ColorUtil
import remix.myplayer.util.DensityUtil
import remix.myplayer.util.SPUtil


/**
 * Created by Remix on 2017/11/22.
 */

class NotifyImpl(context: MusicService) : Notify(context) {

//  private var titleColor: Int? = null
//  private var textColor: Int? = null

//  init {
//    try {
//      val linearLayout = LayoutInflater.from(context).inflate(R.layout.notification_big, null)
//
//      val titleView = linearLayout.findViewById<TextView>(R.id.notify_song)
//      val textView = linearLayout.findViewById<TextView>(R.id.notify_artist_album)
//
//      titleColor = titleView.textColors.defaultColor
//      textColor = textView.textColors.defaultColor
//    } catch (e: Exception) {
//      Timber.w(e)
//    }
//  }


  override fun updateForPlaying() {
    val remoteView = RemoteViews(service.packageName, R.layout.notification)
    val remoteBigView = RemoteViews(service.packageName, R.layout.notification_big)
    val isPlay = service.isPlaying
    val song = service.currentSong
    val isSystemColor = SPUtil.getValue(service, SPUtil.SETTING_KEY.NAME, SPUtil.SETTING_KEY.NOTIFY_SYSTEM_COLOR, true)

    buildAction(service, remoteView, remoteBigView)
    val notification = buildNotification(service, remoteView, remoteBigView)

    //设置歌手，歌曲名
    remoteBigView.setTextViewText(R.id.notify_song, song.title)
    remoteBigView.setTextViewText(R.id.notify_artist_album, song.artist + " - " + song.album)

    remoteView.setTextViewText(R.id.notify_song, song.title)
    remoteView.setTextViewText(R.id.notify_artist_album, song.artist + " - " + song.album)

    //非系统背景色 即黑色背景
    if (!isSystemColor) {
      //字体颜色
      remoteBigView.setTextColor(R.id.notify_song, ColorUtil.getColor(R.color.dark_text_color_primary))
      remoteView.setTextColor(R.id.notify_song, ColorUtil.getColor(R.color.dark_text_color_primary))
      remoteBigView.setTextColor(R.id.notify_artist_album, ColorUtil.getColor(R.color.dark_text_color_secondary))
      remoteView.setTextColor(R.id.notify_artist_album, ColorUtil.getColor(R.color.dark_text_color_secondary))
      //背景
      remoteBigView.setImageViewResource(R.id.notify_bg, R.drawable.bg_notification_black)
      remoteBigView.setViewVisibility(R.id.notify_bg, View.VISIBLE)
      remoteView.setImageViewResource(R.id.notify_bg, R.drawable.bg_notification_black)
      remoteView.setViewVisibility(R.id.notify_bg, View.VISIBLE)
    }

    //桌面歌词
    remoteBigView.setImageViewResource(R.id.notify_lyric,
        if (service.isDesktopLyricLocked) R.drawable.icon_notify_desktop_lyric_unlock else R.drawable.icon_notify_lyric)
    remoteView.setImageViewResource(R.id.notify_lyric,
        if (service.isDesktopLyricLocked) R.drawable.icon_notify_desktop_lyric_unlock else R.drawable.icon_notify_lyric)

    //设置播放按钮
    if (!isPlay) {
      remoteBigView.setImageViewResource(R.id.notify_play, R.drawable.icon_notify_play)
      remoteView.setImageViewResource(R.id.notify_play, R.drawable.icon_notify_play)
    } else {
      remoteBigView.setImageViewResource(R.id.notify_play, R.drawable.icon_notify_pause)
      remoteView.setImageViewResource(R.id.notify_play, R.drawable.icon_notify_pause)
    }

    //设置封面
    val size = DensityUtil.dip2px(service, 128f)
    pushNotify(notification)

    GlideApp.with(service).clear(target)
    target = GlideApp.with(service)
      .asBitmap()
      .load(song)
      .override(size, size)
      .centerCrop()
      .signature(ObjectKey(UriFetcher.albumVersion))
      .into(object : CustomTarget<Bitmap>() {
        override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
          if (song.id == service.currentSong.id) {
            if (!resource.isRecycled) {
              remoteBigView.setImageViewBitmap(R.id.notify_image, resource)
              remoteView.setImageViewBitmap(R.id.notify_image, resource)
            } else {
              remoteBigView.setImageViewResource(R.id.notify_image, R.drawable.album_empty_bg_day)
              remoteView.setImageViewResource(R.id.notify_image, R.drawable.album_empty_bg_day)
            }

            pushNotify(notification)
          }
        }

        override fun onLoadCleared(placeholder: Drawable?) {
        }

        override fun onLoadFailed(errorDrawable: Drawable?) {
          if (song.id == service.currentSong.id) {
            remoteBigView.setImageViewResource(R.id.notify_image, R.drawable.album_empty_bg_day)
            remoteView.setImageViewResource(R.id.notify_image, R.drawable.album_empty_bg_day)
            pushNotify(notification)
          }
        }

        override fun onLoadStarted(placeholder: Drawable?) {
          if (song.id == service.currentSong.id) {
            remoteBigView.setImageViewResource(R.id.notify_image, R.drawable.album_empty_bg_day)
            remoteView.setImageViewResource(R.id.notify_image, R.drawable.album_empty_bg_day)
            pushNotify(notification)
          }
        }
      })
  }

  private fun buildNotification(context: Context, remoteView: RemoteViews, remoteBigView: RemoteViews): Notification {
    val builder = NotificationCompat.Builder(context, PLAYING_NOTIFICATION_CHANNEL_ID)
    builder.setContent(remoteView)
        .setCustomBigContentView(remoteBigView)
        .setContentText("")
        .setContentTitle("")
        .setShowWhen(false)
        .setPriority(NotificationCompat.PRIORITY_MAX)
        .setOngoing(service.isPlaying)
        .setContentIntent(contentIntent)
        .setSmallIcon(R.drawable.icon_notifbar)
    builder.setCustomBigContentView(remoteBigView)
    builder.setCustomContentView(remoteView)
    return builder.build()
  }

  private fun buildAction(context: Context, remoteView: RemoteViews, remoteBigView: RemoteViews) {
    //添加Action
    //切换
    val playIntent = buildPendingIntent(context, Command.TOGGLE)
    remoteBigView.setOnClickPendingIntent(R.id.notify_play, playIntent)
    remoteView.setOnClickPendingIntent(R.id.notify_play, playIntent)
    //下一首
    val nextIntent = buildPendingIntent(context, Command.NEXT)
    remoteBigView.setOnClickPendingIntent(R.id.notify_next, nextIntent)
    remoteView.setOnClickPendingIntent(R.id.notify_next, nextIntent)
    //上一首
    val prevIntent = buildPendingIntent(context, Command.PREV)
    remoteBigView.setOnClickPendingIntent(R.id.notify_prev, prevIntent)

    //关闭通知栏
    val closeIntent = buildPendingIntent(context, Command.CLOSE_NOTIFY)
    remoteBigView.setOnClickPendingIntent(R.id.notify_close, closeIntent)
    remoteView.setOnClickPendingIntent(R.id.notify_close, closeIntent)

    //桌面歌词
    val lyricIntent = buildPendingIntent(context,
        if (service.isDesktopLyricLocked) Command.UNLOCK_DESKTOP_LYRIC else Command.TOGGLE_DESKTOP_LYRIC)
    remoteBigView.setOnClickPendingIntent(R.id.notify_lyric, lyricIntent)
    remoteView.setOnClickPendingIntent(R.id.notify_lyric, lyricIntent)
  }

//  private fun tintBitmap(resId: Int, color: Int?): Bitmap {
//    val bitmap = BitmapFactory.decodeResource(service.resources, resId)
//    return replaceBitmapColor(bitmap, color ?: return bitmap)
//  }
//
//  private fun replaceBitmapColor(oldBitmap: Bitmap, newColor: Int): Bitmap {
//    val copy = oldBitmap.copy(Bitmap.Config.ARGB_8888, true)
//    //循环获得bitmap所有像素点
//    val width = copy.width
//    val height = copy.height
//    for (i in 0 until height) {
//      for (j in 0 until width) {
//        val color = copy.getPixel(j, i)
//        if (color != Color.TRANSPARENT) {
//          copy.setPixel(j, i, newColor)  //替换
//        }
//      }
//    }
//    return copy
//  }

}
